WMF Framework/Significant Events Endpoint/SignificantEventsModels.swift

```diff
@@ -0,0 +1,616 @@
+
+import Foundation
+enum SignificantEventsDecodeError: Error {
+    case unableToParseIntoTypedEvents
+}
+
+public struct SignificantEvents: Decodable {
+    public let nextRvStartId: UInt?
+    public let sha: String?
+    private let untypedEvents: [UntypedEvent]
+    public let typedEvents: [TypedEvent]
+    public let summary: Summary
+    
+    enum CodingKeys: String, CodingKey {
+            case nextRvStartId
+            case sha
+            case untypedEvents = "timeline"
+            case typedEvents
+            case summary
+    }
+    
+    public struct Summary: Decodable {
+        public let earliestTimestampString: String
+        public let numChanges: UInt
+        public let numUsers: UInt
+        
+        enum CodingKeys: String, CodingKey {
+                case earliestTimestampString = "earliestTimestamp"
+                case numChanges
+                case numUsers
+            }
+    }
+    
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        nextRvStartId = try? container.decode(UInt.self, forKey: .nextRvStartId)
+        sha = try? container.decode(String.self, forKey: .sha)
+        summary = try container.decode(Summary.self, forKey: .summary)
+        untypedEvents = try container.decode([UntypedEvent].self, forKey: .untypedEvents)
+        
+        var typedEvents: [TypedEvent] = []
+        
+        for untypedEvent in untypedEvents {
+            switch untypedEvent.outputType {
+            case .small:
+                if let event = Event.Small(untypedEvent: untypedEvent) {
+                    typedEvents.append(.small(event))
+                }
+            case .large:
+                if let event = Event.Large(untypedEvent: untypedEvent) {
+                    typedEvents.append(.large(event))
+                }
+            case .vandalismRevert:
+                if let event = Event.VandalismRevert(untypedEvent: untypedEvent) {
+                    typedEvents.append(.vandalismRevert(event))
+                }
+            case .newTalkPageTopic:
+                if let event = Event.NewTalkPageTopic(untypedEvent: untypedEvent) {
```

Comment:
The test article I've typically been primarily using is COVID-19 Pandemic (https://mobileapps-ios-experiments.wmflabs.org/en.wikipedia.org/v1/page/significant-events/COVID-19_pandemic). Not sure if it's related to any endpoint changes made on Friday, but it looks like there's a parsing issue here that prevents the significant changes for that page from displaying.

1. The `if let event = Event.NewTalkPageTopic(untypedEvent: untypedEvent)` here produces no value because `NewTalkPageTopic`'s initializer https://github.com/wikimedia/wikipedia-ios/blob/917c7335e957200109033f91a35be349b0024ca8/WMF%20Framework/Significant%20Events%20Endpoint/SignificantEventsModels.swift#L232 expects `snippet` to be non-nil. Take a look at the data below to see what was causing the failure for me.

2. This causes the `guard typedEvents.count == untypedEvents.count` statement to fail with false and throws the `unableToParseIntoTypedEvents` error.


```
▿ WMF.SignificantEvents.UntypedEvent
  - outputType: WMF.SignificantEvents.EventOutputType.newTalkPageTopic
  ▿ revId: Optional(981925194)
	- some: 981925194
  ▿ timestampString: Optional("2020-10-05T05:57:47Z")
	- some: "2020-10-05T05:57:47Z"
  ▿ user: Optional("Investigatory")
	- some: "Investigatory"
  ▿ userId: Optional(40140363)
	- some: 40140363
  ▿ userGroups: Optional(["*", "user", "autoconfirmed"])
	▿ some: 3 elements
	  - "*"
	  - "user"
	  - "autoconfirmed"
  ▿ userEditCount: Optional(297)
	- some: 297
  - count: nil
  - sections: nil
  ▿ section: Optional("== RFC to update and confirm current standing of transmission in lead ==")
	- some: "== RFC to update and confirm current standing of transmission in lead =="
  - snippet: nil
  - untypedChanges: nil
```

Reply:
@staykids thanks for the heads up! yeah it seems to me like talk page snippets *shouldn't* return null, but I'll dig into the endpoint to see if it's justified in this case. If so we may have to flag the snippet as optional and hide it in the cell.

Reply:
@staykids this should be fixed in the endpoint now.

Reply:
💪 working great now

Reply:
Hmm seems like this popped up for me again on a different article: https://mobileapps-ios-experiments.wmflabs.org/en.wikipedia.org/v1/page/significant-events/Proud_Boys

In addition to the `snippet` being nil, this one's `section` is also nil - both with non-nil requirements in the `NewTalkPageTopic` initializer.

```
▿ WMF.SignificantEvents.UntypedEvent
  - outputType: WMF.SignificantEvents.EventOutputType.newTalkPageTopic
  ▿ revId: Optional(982152939)
    - some: 982152939
  ▿ timestampString: Optional("2020-10-06T13:18:38Z")
    - some: "2020-10-06T13:18:38Z"
  ▿ user: Optional("ItsPugle")
    - some: "ItsPugle"
  ▿ userId: Optional(30023517)
    - some: 30023517
  ▿ userGroups: Optional(["extendedconfirmed", "*", "user", "autoconfirmed"])
    ▿ some: 4 elements
      - "extendedconfirmed"
      - "*"
      - "user"
      - "autoconfirmed"
  ▿ userEditCount: Optional(3201)
    - some: 3201
  - count: nil
  - sections: nil
  - section: nil
  - snippet: nil
  - untypedChanges: nil
```

Reply:
@staykids thanks for the heads up, I'll look at this next

Reply:
@staykids this should be fixed up better in the endpoint now, it was a side-effect of my previous fix. Let me know if you see any other issues.

Reply:
Nice – looks like an `UntypedEvent` with that revision ID is now being classified as a small change and no longer going through the `NewTalkPageTopic` path. Also just FYI, I'll stick to using test articles from Shay's spreadsheet from here on in – just so happened this article was already on there.